src/write_prometheus.c

```diff
@@ -911,65 +391,110 @@ static int prom_init() {
+      qsort(prom_fam->metric.ptr, prom_fam->metric.num,
+            sizeof(*prom_fam->metric.ptr), prom_metric_cmp);
       continue;
     }
+
+    mmatch->value = m->value;
+
+    /* Prometheus has a globally configured timeout after which metrics are
+     * considered stale. This causes problems when metrics have an interval
+     * exceeding that limit. We emulate the behavior of "pushgateway" and *not*
+     * send a timestamp value – Prometheus will fill in the current time. */
+    if (m->interval > staleness_delta) {
+      static c_complain_t long_metric = C_COMPLAIN_INIT_STATIC;
+      c_complain(LOG_NOTICE, &long_metric,
+                 "write_prometheus plugin: You have metrics with an interval "
+                 "exceeding \"StalenessDelta\" setting (%.3fs). This is "
+                 "suboptimal, please check the collectd.conf(5) manual page to "
+                 "understand what's going on.",
+                 CDTIME_T_TO_DOUBLE(staleness_delta));
+
+      mmatch->time = 0;
+    } else {
+      mmatch->time = m->time;
+    }
   }
 
-  pthread_mutex_unlock(&metrics_lock);
+  pthread_mutex_unlock(&prom_metrics_lock);
   return 0;
 }
 
-static int prom_missing(value_list_t const *vl,
+static int prom_missing(metric_family_t const *fam,
                         __attribute__((unused)) user_data_t *ud) {
-  data_set_t const *ds = plugin_get_ds(vl->type);
-  if (ds == NULL)
-    return ENOENT;
 
-  pthread_mutex_lock(&metrics_lock);
+  pthread_mutex_lock(&prom_metrics_lock);
+
+  metric_family_t *prom_fam = NULL;
+  if (c_avl_get(prom_metrics, fam->name, (void *)&prom_fam) != 0)
+    return 0;
 
-  for (size_t i = 0; i < ds->ds_num; i++) {
-    Io__Prometheus__Client__MetricFamily *fam =
-        metric_family_get(ds, vl, i, /* allocate = */ false);
-    if (fam == NULL)
+  for (size_t i = 0; i < fam->metric.num; i++) {
+    metric_t const *m = &fam->metric.ptr[i];
+
+    metric_t *mmatch = bsearch(m, prom_fam->metric.ptr, prom_fam->metric.num,
+                               sizeof(*prom_fam->metric.ptr), prom_metric_cmp);
+    if (mmatch == NULL)
       continue;
 
-    int status = metric_family_delete_metric(fam, vl);
+    int status = prom_metric_family_metric_delete(prom_fam, m);
     if (status != 0) {
       ERROR("write_prometheus plugin: Deleting a metric in family \"%s\" "
             "failed with status %d",
             fam->name, status);
-
       continue;
     }
 
-    if (fam->n_metric == 0) {
-      int status = c_avl_remove(metrics, fam->name, NULL, NULL);
+    if (prom_fam->metric.num == 0) {
+      int status = c_avl_remove(prom_metrics, prom_fam->name, NULL, NULL);
       if (status != 0) {
         ERROR("write_prometheus plugin: Deleting metric family \"%s\" failed "
               "with status %d",
-              fam->name, status);
+              prom_fam->name, status);
         continue;
       }
-      metric_family_destroy(fam);
+      metric_family_free(prom_fam);
```

Comment:
After freeing "prom_fam", you cannot loop & use "prom_fam" members any more => you should break out of the loop here.